keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,623 @@
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The new structure introduces a configurable timeout that allows administrators to
+define a timeout for webhooks in case they are unreachable.
+
+> Note: ABAC will not be supported by the new file format as we discourage the usage of this mode.
+
+The code path for enabling the above will only be triggered if the feature flag will
+be enabled until the time the feature flag is removed and configuring authorizer
+through a file becomes GA.
+
+> TODO: More details on the implementation will be charted once the goals, non-goal and proposal is approved.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+- `mode` {RBAC, Node, Webhook}
+- `decision` {Allow, Deny, NoOpinion}
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track
+total invocation counts across the following labels.
+
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+3. `apiserver_authorization_step_webhook_duration_seconds`
+
+This metric will track the average latency
+
+Labels {along with possible values}:
+- `server`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+4. `apiserver_authorization_step_webhook_error_total`
+
+This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.
+
+Labels {along with possible values}:
+
+- server
+- `code` {4xx, 5xx}
+- `decision` {Deny, NoOpinion}
+- `subject_access_review_version` {v1beta1, v1}
+
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests would be added along with any new code introduced.
+
+Existing test coverage:
+- cmd/kube-apiserver/app/options: 65
+- staging/src/k8s.io/apiserver/pkg/authorization: NA
+- staging/src/k8s.io/apiserver/apis/config/v1: 54.5
+
+> Will add more packages to the list during implementation.
+
+##### Integration tests
+
+Integration tests would be added to ensure the following:
+- Authorization of requests work in the existing flag based mode (feature flag
+turned off)
+- Authorization of requests work with an apiserver bootstrapped with
+authorization configuration file (feature flag turned on)
+    - without an webhook
+    - with an webhook - successful request
+    - with an webhook - error on request with `onError: Deny`
+    - with an webhook - error on request with `onError: NoOpinion`
```

Comment:
You needs tests that mix and match the different authorizers in different orders.